File: Core/src/Internal/OftPeer.cs
namespace BlueHeighliner.OpenFrameTransport.Internal;

/// <summary>
/// <inheritdoc cref="IOftPeer" />
/// </summary>
internal sealed class OftPeer : IOftPeer
{
    private readonly OftPeerOptions options;
    private readonly IOftConnector connector;
    private readonly OftConnectionOptions connectOptions;
    private readonly IOftHoster hoster;
    private readonly OftConnectionOptions hostOptions;

    private readonly Dictionary<(string Host, int Port), Task<IOftConnection>> outboundConnections = new();
    private readonly object outboundLock = new();

    private readonly List<IOftConnection> inboundConnections = [];
    private readonly object inboundLock = new();

    private readonly OftBufferedHandlerSlot<Action<IOftPeerReception>> receivedSlot = new();

    /// <summary>
    /// How long a connection must have had no pending data (see
    /// <see cref="IOftConnection.HasPendingData"/>) before it becomes eligible for automatic
    /// eviction (idle, lifetime, or excess-count based) at all — a fixed value, not configurable,
    /// giving the underlying TLS/TCP layers time to actually flush and acknowledge everything after
    /// the last application-level message completes, rather than evicting the instant
    /// <see cref="IOftConnection.HasPendingData"/> turns <see langword="false"/>.
    /// </summary>
    private static readonly TimeSpan EvictionGracePeriod = TimeSpan.FromSeconds(30);

    /// <summary>
    /// How often <see cref="RunEviction"/> checks connections against
    /// <see cref="OftPeerOptions.IdleTimeout"/>, <see cref="OftPeerOptions.MaxConnectionLifetime"/>,
    /// and <see cref="OftPeerOptions.MaxConnectionCount"/> — a fixed value, not configurable (see
    /// <see cref="IOftPeer"/>'s own doc comment). Since eviction only ever runs on this cadence,
    /// neither of those two duration-based options can take effect any sooner than this floor.
    /// </summary>
    private static readonly TimeSpan EvictionCheckInterval = TimeSpan.FromSeconds(30);

    /// <summary>
    /// When each tracked connection was first observed, by <see cref="RunEviction"/>, to have no
    /// pending data (see <see cref="IOftConnection.HasPendingData"/>) — cleared again the moment a
    /// connection is next observed with pending data, so a connection that resumes sending after a
    /// quiet period gets a fresh <see cref="EvictionGracePeriod"/> once it finishes again. A
    /// connection isn't a candidate for any automatic eviction (idle, lifetime, or excess-count)
    /// until this has aged past <see cref="EvictionGracePeriod"/>.
    /// </summary>
    private readonly Dictionary<IOftConnection, DateTimeOffset> pendingDataClearedAt = new();
    private readonly object pendingDataClearedAtLock = new();

    private IOftListener? listener;
    private readonly object listenerLock = new();

    private readonly Timer evictionTimer;

    private readonly object disconnectLock = new();
    private bool disconnected;

    /// <summary>
    /// Creates a peer using the given options and the connector/hoster it delegates to. Constructed
    /// by <see cref="OftPeerFactory"/>, which builds <paramref name="connectOptions"/>/
    /// <paramref name="hostOptions"/> from an <see cref="OftPeerOptions"/>.
    /// </summary>
    /// <param name="options">The peer's options.</param>
    /// <param name="connector">The connector used to make outbound connections.</param>
    /// <param name="connectOptions">The options used for every outbound connection this peer makes.</param>
    /// <param name="hoster">The hoster used to accept inbound connections.</param>
    /// <param name="hostOptions">The options used to accept every inbound connection.</param>
    public OftPeer(OftPeerOptions options, IOftConnector connector, OftConnectionOptions connectOptions, IOftHoster hoster, OftConnectionOptions hostOptions)
    {
        this.options = options;
        this.connector = connector;
        this.connectOptions = connectOptions;
        this.hoster = hoster;
        this.hostOptions = hostOptions;

        this.evictionTimer = new Timer(_ => this.RunEviction(), null, EvictionCheckInterval, EvictionCheckInterval);
    }

    /// <inheritdoc />
    public bool IsConnected => !this.disconnected;

    /// <inheritdoc />
    public IPEndPoint? LocalEndPoint
    {
        get
        {
            lock (this.listenerLock)
            {
                return this.listener?.LocalEndPoint;
            }
        }
    }

    /// <inheritdoc />
    public Action<IOftPeerReception>? ReceivedHandler
    {
        get => this.receivedSlot.Handler;
        set => this.receivedSlot.Handler = value;
    }

    /// <inheritdoc />
    public async Task Listen(IPEndPoint listenEndPoint, CancellationToken cancellationToken = default)
    {
        ObjectDisposedException.ThrowIf(this.disconnected, this);

        IOftListener opened = await this.hoster.Host(listenEndPoint, this.hostOptions, cancellationToken).ConfigureAwait(false);
        opened.ConnectedHandler = connection =>
        {
            lock (this.inboundLock)
            {
                this.inboundConnections.Add(connection);
            }

            this.TrackConnection(connection, tracked =>
            {
                lock (this.inboundLock)
                {
                    this.inboundConnections.Remove(tracked);
                }
            });
        };

        lock (this.listenerLock)
        {
            this.listener = opened;
        }
    }

    /// <inheritdoc />
    public Task StopListening()
    {
        ObjectDisposedException.ThrowIf(this.disconnected, this);

        IOftListener? currentListener;
        lock (this.listenerLock)
        {
            currentListener = this.listener;
            this.listener = null;
        }

        currentListener?.Dispose();
        return Task.CompletedTask;
    }

    /// <inheritdoc />
    public async Task Send(string host, int port, ReadOnlyMemory<byte> data, int priority = 0, CancellationToken cancellationToken = default)
    {
        if (this.disconnected)
        {
            throw new OftDisconnectedException("This peer is no longer connected.");
        }

        IOftConnection connection = await this.GetOrCreateConnection(host, port, cancellationToken).ConfigureAwait(false);
        await connection.Send(data, priority, cancellationToken).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task Send(string host, int port, IMemoryOwner<byte> data, int priority = 0, CancellationToken cancellationToken = default)
    {
        if (this.disconnected)
        {
            throw new OftDisconnectedException("This peer is no longer connected.");
        }

        IOftConnection connection = await this.GetOrCreateConnection(host, port, cancellationToken).ConfigureAwait(false);
        await connection.Send(data, priority, cancellationToken).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public Task Rekey(CancellationToken cancellationToken = default)
    {
        if (this.disconnected)
        {
            throw new OftDisconnectedException("This peer is no longer connected.");
        }

        return Task.WhenAll(this.GetTrackedConnections().Select(connection => connection.Rekey(cancellationToken)));
    }

    /// <inheritdoc />
    public Task Drop()
    {
        ObjectDisposedException.ThrowIf(this.disconnected, this);
        return Task.WhenAll(this.GetTrackedConnections().Select(connection => connection.Disconnect()));
    }

    /// <inheritdoc />
    public async Task Disconnect()
    {
        if (!this.TryBeginDisconnect())
        {
            return;
        }

        this.evictionTimer.Dispose();

        IOftListener? currentListener;
        lock (this.listenerLock)
        {
            currentListener = this.listener;
            this.listener = null;
        }

        currentListener?.Dispose();

        await Task.WhenAll(this.GetTrackedConnections().Select(connection => connection.Disconnect())).ConfigureAwait(false);

        this.receivedSlot.DisposeBuffered();
    }

    /// <summary>
    /// Immediately and synchronously puts this peer into a disconnected state: stops listening (if
    /// applicable), immediately terminates every connection it currently holds without waiting for
    /// any of their background work to finish, and releases every other resource it owns. Call
    /// <see cref="Disconnect"/> instead for a graceful, awaitable teardown.
    /// </summary>
    public void Dispose()
    {
        if (!this.TryBeginDisconnect())
        {
            return;
        }

        this.evictionTimer.Dispose();

        IOftListener? currentListener;
        lock (this.listenerLock)
        {
            currentListener = this.listener;
            this.listener = null;
        }

        currentListener?.Dispose();

        foreach (IOftConnection connection in this.GetTrackedConnections())
        {
            connection.Dispose();
        }

        this.receivedSlot.DisposeBuffered();
    }

    /// <summary>
    /// Atomically transitions this peer into a disconnected state exactly once: returns
    /// <see langword="true"/> the first time this is called (by either <see cref="Disconnect"/> or
    /// <see cref="Dispose"/>), and <see langword="false"/> every time after, for both of them - so
    /// whichever is called first performs the actual teardown, and any subsequent call to either is
    /// a no-op.
    /// </summary>
    private bool TryBeginDisconnect()
    {
        lock (this.disconnectLock)
        {
            if (this.disconnected)
            {
                return false;
            }

            this.disconnected = true;
            return true;
        }
    }

    private Task<IOftConnection> GetOrCreateConnection(string host, int port, CancellationToken cancellationToken)
    {
        (string Host, int Port) key = (host, port);

        Task<IOftConnection>? connectTask;
        lock (this.outboundLock)
        {
            if (!this.outboundConnections.TryGetValue(key, out connectTask))
            {
                connectTask = this.Connect(host, port, cancellationToken);
                this.outboundConnections[key] = connectTask;
            }
        }

        return this.AwaitAndUntrackOnFailure(key, connectTask);
    }

    private async Task<IOftConnection> AwaitAndUntrackOnFailure((string Host, int Port) key, Task<IOftConnection> connectTask)
    {
        try
        {
            return await connectTask.ConfigureAwait(false);
        }
        catch
        {
            lock (this.outboundLock)
            {
                if (this.outboundConnections.TryGetValue(key, out Task<IOftConnection>? current) && current == connectTask)
                {
                    this.outboundConnections.Remove(key);
                }
            }

            throw;
        }
    }

    private async Task<IOftConnection> Connect(string host, int port, CancellationToken cancellationToken)
    {
        // Assigning ReceivedHandler/DisconnectedHandler after this.connector.Connect returns is safe:
        // they're backed by OftBufferedHandlerSlot, so nothing the connection raised in the meantime
        // is lost (see README.md and OftBufferedHandlerSlot's own doc comment).
        IOftConnection connection = await this.connector.Connect(host, port, this.connectOptions, cancellationToken).ConfigureAwait(false);

        this.TrackConnection(connection, tracked =>
        {
            lock (this.outboundLock)
            {
                if (this.outboundConnections.TryGetValue((host, port), out Task<IOftConnection>? current) &&
                    current.IsCompletedSuccessfully && current.Result == tracked)
                {
                    this.outboundConnections.Remove((host, port));
                }
            }
        });

        return connection;
    }

    /// <summary>
    /// Forwards a tracked connection's received messages to this peer's own
    /// <see cref="ReceivedHandler"/>, and runs <paramref name="onDisconnectedTrackingCleanup"/> when
    /// it disconnects to untrack it (from <see cref="outboundConnections"/> or
    /// <see cref="inboundConnections"/> as appropriate) — this peer has no external disconnected
    /// notification of its own to forward to (see <see cref="IOftPeer.ReceivedHandler"/>'s own doc
    /// comment for why).
    /// </summary>
    private void TrackConnection(IOftConnection connection, Action<IOftConnection> onDisconnectedTrackingCleanup)
    {
        connection.ReceivedHandler = data =>
        {
            OftPeerReception reception = new(data, connection.Identity);
            this.receivedSlot.Raise(callback => callback(reception), discardedDisposable: reception);
        };
        connection.DisconnectedHandler = _ =>
        {
            onDisconnectedTrackingCleanup(connection);

            lock (this.pendingDataClearedAtLock)
            {
                this.pendingDataClearedAt.Remove(connection);
            }
        };
    }

    /// <summary>
    /// Every connection this peer currently holds, both outbound (only those that have finished
    /// connecting successfully) and inbound.
    /// </summary>
    private List<IOftConnection> GetTrackedConnections()
    {
        List<IOftConnection> tracked = [];

        lock (this.outboundLock)
        {
            foreach (Task<IOftConnection> connectTask in this.outboundConnections.Values)
            {
                if (connectTask.IsCompletedSuccessfully)
                {
                    tracked.Add(connectTask.Result);
                }
            }
        }

        lock (this.inboundLock)
        {
            tracked.AddRange(this.inboundConnections);
        }

        return tracked;
    }

    private void RunEviction()
    {
        DateTimeOffset now = DateTimeOffset.UtcNow;
        List<IOftConnection> tracked = this.GetTrackedConnections();

        // A connection with pending/unacknowledged data (see IOftConnection.HasPendingData) is
        // never auto-disconnected here, regardless of which eviction condition it would otherwise
        // meet: doing so could silently drop a message that's still queued, in flight, or only
        // partially reassembled. It's only a candidate once all of its data has been acknowledged,
        // and even then, only once EvictionGracePeriod has passed since that happened - giving the
        // underlying TLS/TCP layers time to actually flush and acknowledge everything rather than
        // evicting the instant HasPendingData turns false.
        List<IOftConnection> evictionCandidates = [];
        lock (this.pendingDataClearedAtLock)
        {
            foreach (IOftConnection connection in tracked)
            {
                if (connection.HasPendingData)
                {
                    this.pendingDataClearedAt.Remove(connection);
                    continue;
                }

                if (!this.pendingDataClearedAt.TryGetValue(connection, out DateTimeOffset clearedAt))
                {
                    clearedAt = now;
                    this.pendingDataClearedAt[connection] = clearedAt;
                }

                if (now - clearedAt >= EvictionGracePeriod)
                {
                    evictionCandidates.Add(connection);
                }
            }
        }

        HashSet<IOftConnection> toDisconnect = [];
        foreach (IOftConnection connection in evictionCandidates)
        {
            DateTimeOffset lastActivity = connection.LastSentAt > connection.LastReceivedAt ? connection.LastSentAt : connection.LastReceivedAt;
            if (now - lastActivity > this.options.IdleTimeout || now - connection.ConnectedAt > this.options.MaxConnectionLifetime)
            {
                toDisconnect.Add(connection);
            }
        }

        int remainingCount = tracked.Count - toDisconnect.Count;
        if (remainingCount > this.options.MaxConnectionCount)
        {
            int excess = remainingCount - this.options.MaxConnectionCount;
            foreach (IOftConnection connection in evictionCandidates
                .Where(connection => !toDisconnect.Contains(connection))
                .OrderBy(connection => connection.ConnectedAt)
                .Take(excess))
            {
                toDisconnect.Add(connection);
            }
        }

        foreach (IOftConnection connection in toDisconnect)
        {
            _ = connection.Disconnect();
        }
    }
}

File: Ports/Java/src/main/java/org/blueheighliner/openframetransport/OftPeer.java
package org.blueheighliner.openframetransport;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;

/**
 * A peer-to-peer convenience layer over {@link OftHoster}/{@link OftListener} and
 * {@link OftConnector}. Sending a message to a host/port transparently reuses an existing
 * connection or creates and caches a new one; idle, expired, or excess cached connections are
 * disconnected automatically, and connections with a configured
 * {@link OftPeerOptions#rekeyInterval()} rekey themselves automatically (see README.md &sect;8).
 * A connection only ever becomes eligible for automatic disconnection once it has had no pending
 * data (see {@link OftConnection#hasPendingData()}) for a fixed 30-second grace period - not
 * configurable - giving the underlying TLS/TCP layers time to actually flush and acknowledge
 * everything after the last application-level message completes. Eviction itself (checking
 * connections against {@link OftPeerOptions#idleTimeout()}, {@link OftPeerOptions#maxConnectionLifetime()},
 * and {@link OftPeerOptions#maxConnectionCount()}) is likewise only ever run on a fixed,
 * non-configurable 30-second interval - so, combined with the grace period above, neither
 * {@link OftPeerOptions#idleTimeout()} nor {@link OftPeerOptions#maxConnectionLifetime()} can take
 * effect any sooner than roughly 30-60 seconds after it's reached, regardless of how much shorter
 * either is configured to be. There is no way to enumerate or
 * look up an individual connection this peer holds;
 * {@link #rekey()} and {@link #drop()} act on all of them at once. {@link #close()} permanently
 * puts this peer itself into a disconnected state - unlike {@link #drop()}, which only disconnects
 * this peer's currently held connections and leaves the peer itself usable - after which
 * {@link #isConnected()} is permanently {@code false} and every other member below throws:
 * {@link #listen}, {@link #stopListening()}, and {@link #drop()} throw {@link IllegalStateException}, while
 * {@link #send} and {@link #rekey()} throw {@link OftDisconnectedException}.
 */
public interface OftPeer extends AutoCloseable {
    /**
     * Creates a peer using the given options, building the connector and hoster it delegates to.
     *
     * @param options the peer's options
     * @return the new peer
     * @throws IllegalArgumentException {@code options.securityMode()} is
     *                                  {@link OftSecurityMode#SERVER_AUTHENTICATION} — not a valid
     *                                  mode for a peer, which has no client/server delineation and
     *                                  so cannot express a one-sided authentication requirement (use
     *                                  {@link OftSecurityMode#DUAL_AUTHENTICATION} instead)
     */
    static OftPeer create(OftPeerOptions options) {
        if (options.securityMode() == OftSecurityMode.SERVER_AUTHENTICATION) {
            throw new IllegalArgumentException(
                    "SERVER_AUTHENTICATION is not a valid securityMode for an OftPeer: a peer has no client/server "
                            + "delineation, so it cannot express a one-sided authentication requirement. Use "
                            + "DUAL_AUTHENTICATION instead.");
        }

        return new DefaultOftPeer(options);
    }

    /**
     * Whether this peer is still connected: {@code true} until {@link #close()} is called, after
     * which it is permanently {@code false}. Unlike {@link OftConnection#isConnected()}, this is
     * unaffected by any individual connection this peer holds disconnecting (locally via
     * {@link #drop()} or remotely) - connection lifecycle is this peer's own implementation detail
     * (see {@link #setReceivedHandler}'s own doc comment).
     */
    boolean isConnected();

    /**
     * The endpoint actually being listened on once {@link #listen(InetSocketAddress)} has completed,
     * or {@code null} if the peer isn't currently listening.
     */
    InetSocketAddress getLocalEndpoint();

    /**
     * Called for every message received on any connection this peer holds, both inbound and
     * outbound, with the identity of the connection it arrived on and its payload, or {@code null} if
     * no callback is currently assigned. There is only ever one callback at a time - assigning a new
     * value here always replaces any previous one, and the same buffering-until-first-non-null-assignment
     * guarantee {@link OftConnection#setReceivedHandler} itself makes applies here too (see
     * README.md). This peer deliberately exposes no way to enumerate, look up, or be notified about
     * the individual connections it holds beyond the identity passed here (e.g. no disconnected
     * notification): connection lifecycle is this peer's own implementation detail, transparently
     * managed (reconnecting, evicting, etc.) behind {@link #send}.
     */
    void setReceivedHandler(BiConsumer<OftIdentity, byte[]> handler);

    /** The callback currently assigned via {@link #setReceivedHandler}, or {@code null} if none is. */
    BiConsumer<OftIdentity, byte[]> getReceivedHandler();

    /**
     * Called whenever a message sent with a non-null {@code tag} (see {@link #send}) has been fully
     * delivered and acknowledged (see {@link OftConnection#setAcknowledgedHandler}), with the identity
     * of the connection it was sent over and that same tag. Never called for a message sent with a
     * {@code null} tag, or for one that was cancelled rather than delivered. {@code null} if no
     * callback is currently assigned. There is only ever one callback at a time - assigning a new
     * value here always replaces any previous one. Unlike {@link #setReceivedHandler}, this does
     * <em>not</em> buffer a raise that happens before a callback is ever assigned - see
     * {@link OftConnection#setAcknowledgedHandler}'s own doc comment for why that's safe.
     */
    void setAcknowledgedHandler(BiConsumer<OftIdentity, Object> handler);

    /** The callback currently assigned via {@link #setAcknowledgedHandler}, or {@code null} if none is. */
    BiConsumer<OftIdentity, Object> getAcknowledgedHandler();

    /**
     * Starts listening for inbound connections. A peer that never calls this only ever makes
     * outbound connections. Binding runs on a dedicated background thread (see
     * {@link OftHoster#host(InetSocketAddress, OftHostOptions)}), not the calling thread.
     *
     * @param listenEndpoint the local endpoint to listen for incoming TCP connections on
     * @return a future that completes once this peer is listening, or completes exceptionally if
     * binding fails
     * @throws IllegalStateException {@link #isConnected()} is {@code false} - thrown synchronously
     *                                rather than deferred to the returned future, since it's an
     *                                immediate state check rather than something that requires
     *                                binding
     */
    CompletableFuture<Void> listen(InetSocketAddress listenEndpoint);

    /**
     * Stops listening for new inbound connections. Already-established connections are left open.
     * Not named {@code close()}, matching C#'s {@code IOftPeer.StopListening()} and C's
     * {@code oft_peer_stop_listening()}: that name is reserved here, as in both, for
     * {@link AutoCloseable#close()}'s full-teardown semantics (see {@link #close()}).
     *
     * @return a future that completes once this peer has stopped listening
     * @throws IllegalStateException {@link #isConnected()} is {@code false}
     */
    CompletableFuture<Void> stopListening();

    /**
     * Sends a message to {@code host}:{@code port}, reusing a cached connection if one already
     * exists, or creating and caching a new one otherwise.
     *
     * @param host     the remote host to send to
     * @param port     the remote port to send to
     * @param data     the message payload
     * @param priority the priority to send the message at (see README.md &sect;5-&sect;6)
     * @param tag      an opaque, application-controlled value attached to this send, so it can be
     *                 referenced later - passed back to {@link #setAcknowledgedHandler}'s callback,
     *                 along with the identity of the connection it was sent over, once this message is
     *                 fully delivered and acknowledged, if non-null; {@code null} means this send
     *                 never raises it
     * @return a handle that can be used to wait for delivery or cancel the message
     * @throws OftDisconnectedException {@link #isConnected()} is {@code false}
     */
    OftSendHandle send(String host, int port, byte[] data, int priority, Object tag) throws IOException;

    /**
     * Requests a TLS 1.3 {@code KeyUpdate} (see README.md &sect;8) on every connection this peer
     * currently holds, both outbound and inbound. Connections established after this call is
     * issued are unaffected.
     *
     * @return a future that completes once every connection's local key update request has been sent
     * @throws OftDisconnectedException {@link #isConnected()} is {@code false}
     */
    CompletableFuture<Void> rekey();

    /**
     * Disconnects every connection this peer currently holds, both outbound and inbound. Unlike
     * {@link #close()}, this peer itself is left usable afterward - a subsequent {@link #send} call
     * creates and caches a new outbound connection as usual, and, if listening, new inbound
     * connections keep being accepted.
     *
     * @return a future that completes once every held connection's teardown has been requested
     * @throws IllegalStateException {@link #isConnected()} is {@code false}
     */
    CompletableFuture<Void> drop();

    /**
     * Permanently puts this peer itself into a disconnected state: stops listening (if applicable)
     * and disconnects every connection it currently holds, both outbound and inbound. Safe to call
     * more than once; every call after the first is a no-op.
     */
    @Override
    void close();
}
